net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+#include <net/inet_ecn.h>
+#include <asm/byteorder.h>
+
+/* Priority markers */
+#define MAX_PRIORITY 63
+#define HALF_MAX_PRIORITY 31
+
+
+
+/*	  Gatekeeper Priority Queue
+ *	=================================
+ *
+ *
+ * Gatekeeper is an open source DoS defense software which sends traffic based   * on an assigned priority for the client's requeust. Assigning priorities to    * packets enables us to give low priority to clients who quickly send many      * requests, so  that we can service legitimate clients (who typically send      * requests more slowly). The DoS scenario is when assigning low priorities      * helps block out malicious senders, who could otherwise overwhelm Gatekeeper   * with a very high number of requests, denying service to the requests of 
+ * legitimate users.
+ *
+ *
+ * To assign priority to packet, Gatekeeper uses the six bits of the DSCP field, * as it is a sufficient and convenient way to limit the maximum number of       * priorities we'd like to deal with. Additionally, using the DSCP field also    * allows us to more easily use tools like iptables and tc, which already        * understand fields in the IP header. 
+ *
+ *
+ * Although the DSCP field is usually used to specify DiffServe traffic 
+ * classifcation options, it is NOT used for that intent here. 
+ *
+ *
+ *
+ *   The Gatekeper standalone codebase is found here: 
+ *	          
+ *          https://github.com/AltraMayor/gatekeeper
+ *
+ *
+ */
+
+
+struct gkprioq_sched_data {
+/* Parameters */
+	u32 max_limit;
+
+/*Variables*/
+	struct Qdisc *prio_array[MAX_PRIORITY + 1];
+	u16 last_dropped_prio;
+	u16 highest_prio;
+	u32 init_limit;
+};
+
+static inline u32 calc_alloc_len(const struct gkprioq_sched_data *q)
+{
+	u32 total_alloc_len = 0;
+	int prio;
+	
+	for (prio = 0; prio < MAX_PRIORITY + 1; prio++) 
+		total_alloc_len += (q->prio_array[prio])->limit;
+
+	return total_alloc_len;
+}
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	if (q->highest_prio > (HALF_MAX_PRIORITY + 1)) {
+		for (prio = q->highest_prio - 1; prio > HALF_MAX_PRIORITY + 1; 
+		                                 prio--) {
+			if ((q->prio_array[prio])->q.qlen > 0)
+				return prio;
+		}
+	}
+
+	if ((q->prio_array[0])->q.qlen > 0) /*Treating as best effort queue*/
+		return 0;
+
+	for (prio = HALF_MAX_PRIORITY + 1; prio > 0; prio--) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+
+	return 0; /*GK queue is empty, return default highest priority setting*/
+
+}
+
+static int gkprioq_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct Qdisc *your_queue;
+	int prio, ret, err, wlen, gk_len;
+	u16 your_prio; 
+
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen) || 
+		    skb_try_make_writable(skb, wlen))
```

Comment:
Hi Michel, the entire block that pulls the DSCP field from the packet header (the whole switch block), is written directly from sch_dsmark.c just as we discussed in the Gmail thread. I haven't changed any line and it is the same as in sch_dsmark.c (http://elixir.free-electrons.com/linux/latest/source/net/sched/sch_dsmark.c#L203) . Having said that, it does seem to work fine.